internal/v2/application/command.go

```diff
@@ -0,0 +1,740 @@
+	readings := make([]dtos.BaseReading, 0, configuration.Device.MaxCmdOps)
+
+	for _, cv := range cvs {
+		// double check the CommandValue return from ProtocolDriver match device command
+		dr, ok := cache.Profiles().DeviceResource(c.device.Profile.Name, cv.DeviceResourceName)
+		if !ok {
+			return nil, fmt.Errorf("no deviceResource %s for %s in CommandValue (%s)", cv.DeviceResourceName, c.device.Name, cv.String())
+		}
+
+		// perform data transformation
+		if configuration.Device.DataTransform {
+			err = transformer.TransformReadResult(cv, dr.Properties.Value, lc)
+			if err != nil {
+				lc.Error(fmt.Sprintf("failed to transform CommandValue (%s): %v", cv.String(), err))
+				transformsOK = false
+			}
+		}
+
+		// assertion
+		dc := container.MetadataDeviceClientFrom(c.dic.Get)
+		err = transformer.CheckAssertion(cv, dr.Properties.Value.Assertion, c.device, lc, dc)
+		if err != nil {
+			cv = dsModels.NewStringValue(cv.DeviceResourceName, cv.Origin, fmt.Sprintf("Assertion failed for device resource: %s, with value: %s", cv.DeviceResourceName, cv.String()))
+		}
+
+		// ResourceOperation mapping
+		ro, err := cache.Profiles().ResourceOperation(c.device.Profile.Name, cv.DeviceResourceName, sdkCommon.GetCmdMethod)
+		if err != nil {
+			// this allows SDK to directly read deviceResource without deviceCommands defined.
+			lc.Debug(fmt.Sprintf("failed to read ResourceOperation: %v", err))
+		} else if len(ro.Mappings) > 0 {
+			newCV, ok := transformer.MapCommandValue(cv, ro.Mappings)
+			if ok {
+				cv = newCV
+			} else {
+				lc.Warn(fmt.Sprintf("ResourceOperation (%s) mapping value (%s) failed with the mapping table: %v", ro.DeviceResource, cv.String(), ro.Mappings))
+			}
+		}
+
+		reading := commandValueToReading(cv, c.device.Name, dr.Properties.Value.MediaType, dr.Properties.Value.FloatEncoding)
+		readings = append(readings, *reading)
+
+		if cv.Type == dsModels.Binary {
+			lc.Debug(fmt.Sprintf("device: %s DeviceResource: %v reading: binary value", c.device.Name, cv.DeviceResourceName))
+		} else {
+			lc.Debug(fmt.Sprintf("device: %s DeviceResource: %v reading: %v", c.device.Name, cv.DeviceResourceName, reading))
+		}
+	}
+
+	if !transformsOK {
+		return nil, fmt.Errorf("GET command %s transform failed for %s", cmd, c.device.Name)
+	}
+
+	eventDTO := &dtos.Event{DeviceName: c.device.Name, Readings: readings}
+	eventDTO.Origin = sdkCommon.GetUniqueOrigin()
+
+	return eventDTO, nil
+}
+
+func parseParams(params string) (paramMap map[string]string, err error) {
+	err = json.Unmarshal([]byte(params), &paramMap)
+	if err != nil {
+		return
+	}
+
+	if len(paramMap) == 0 {
+		err = fmt.Errorf("no parameters specified")
+		return
+	}
+
+	return
+}
+
+func createCommandValueFromDeviceResource(dr *contract.DeviceResource, v string) (*dsModels.CommandValue, error) {
+	var err error
+	var result *dsModels.CommandValue
+
+	origin := time.Now().UnixNano()
+	switch strings.ToLower(dr.Properties.Value.Type) {
+	case "string":
```

Comment:
use `strings.ToLower(models.ValueTypeString)` from go-mod-core-contracts, and make the same changes for each value type